programs/whirlpool/src/math/token_math.rs

```diff
@@ -84,11 +133,42 @@ pub fn get_amount_delta_b(
     liquidity: u128,
     round_up: bool,
 ) -> Result<u64, ErrorCode> {
-    let (price_lower, price_upper) = increasing_price_order(sqrt_price_0, sqrt_price_1);
+    match try_get_amount_delta_b(sqrt_price_0, sqrt_price_1, liquidity, round_up) {
+        Ok(AmountDeltaU64::Valid(value)) => Ok(value),
+        Ok(AmountDeltaU64::ExceedsMax(error)) => Err(error),
+        Err(error) => Err(error),
+    }
+}
 
-    // liquidity * (price_upper - price_lower) must be less than 2^128
-    // for the token amount to be less than 2^64
-    checked_mul_shift_right_round_up_if(liquidity, price_upper - price_lower, round_up)
+pub fn try_get_amount_delta_b(
+    sqrt_price_0: u128,
+    sqrt_price_1: u128,
+    liquidity: u128,
+    round_up: bool,
+) -> Result<AmountDeltaU64, ErrorCode> {
+    let (sqrt_price_lower, sqrt_price_upper) = increasing_price_order(sqrt_price_0, sqrt_price_1);
+
+    // customized checked_mul_shift_right_round_up_if
```

Comment:
sorry if I'm missing something, why can't this function call into `checked_mul_shift_right_round_up_if` rather than duplicating the logic here?

I think it should be possible to just match on the error codes

Reply:
I changed this from the original draft using `match` because I felt that it may be an anti-pattern for the caller to know about all errors generated by callee functions.

If the error code is changed or the same error code is returned in multiple places in the function, it is difficult for the caller to notice and to identify the difference.